File: lab1/SimUDuck/Duck.cpp
#include "stdafx.h"
#include "Duck.h"

Duck::Duck(std::unique_ptr<IFlyBehavior>&& flyBehavior,
	std::unique_ptr<IQuackBehavior>&& quackBehavior)
	:m_quackBehavior(std::move(quackBehavior))
{
	assert(m_quackBehavior);
	SetFlyBehavior(std::move(flyBehavior));
}

void Duck::Quack() const
{
	m_quackBehavior->Quack();
}

void Duck::Swim()
{
	std::cout << "I'm swimming" << std::endl;
}

void Duck::Fly()
{
	m_flyBehavior->Fly();
}

void Duck::Dance()
{
	std::cout << "I'm Dancing" << std::endl;
}

void Duck::SetFlyBehavior(std::unique_ptr<IFlyBehavior>&& flyBehavior)
{
	assert(flyBehavior);
	m_flyBehavior = std::move(flyBehavior);
}

File: lw1/src/duck/Duck.php
<?php

abstract class Duck
{
    /** @var IFlyBehavior */
    private $flyBehavior;
    /** @var IQuackBehavior */
    private $quackBehavior;
    /** @var IDanceBehavior */
    private $danceBehavior;

    public function __construct(IFlyBehavior $flyBehavior,
                                IQuackBehavior $quackBehavior,
                                IDanceBehavior $danceBehavior)
    {
        $this->flyBehavior = $flyBehavior;
        $this->quackBehavior = $quackBehavior;
        $this->danceBehavior = $danceBehavior;
    }

    public function quack(): void
    {
        $this->quackBehavior->quack();
    }

    public function swim(): void
    {
        echo 'I\'m swimming' . PHP_EOL;
    }

    public function fly(): void
    {
        $this->flyBehavior->fly();
    }

    public function dance(): void
    {
        $this->danceBehavior->dance();
    }

    public function setFlyBehavior(IFlyBehavior $flyBehavior)
    {
        $this->flyBehavior = $flyBehavior;
    }

    abstract public function display(): void;
}